### PR TITLE
basemaps: first modals (xyz)

### PR DESCRIPTION
**Pull request by matallo**
still finishing a detail (add basemap icon) which I'll check with @piensaenpixel but it's ready for review in terms of functionality

basemapsmodals

CR @xavijam

### CODE REVIEW

File: lib/assets/test/spec/cartodb3/editor/layers/basemap-content-views/basemap-select-view.spec.js [hunk 1]
```diff
@@ -157,21 +206,38 @@ describe('editor/layers/basemap-content-views/basemap-select-view', function ()
...
       });
       this.view.render();
 
-      // mapbox go under custom by now
       expect(this.view._filteredBasemapsCollection.length).toBe(1);
-      expect(this.view.$('.Mosaic-item').length).toBe(1);
+      expect(this.view.$('.Mosaic-item').length).toBe(2); // Custom basemap and "Add basemap" button
     });
 
+    // it('should render mosaic if category is Mapbox', function () {
```
Comment (xavijam) [inline lib/assets/test/spec/cartodb3/editor/layers/basemap-content-views/basemap-select-view.spec.js]: Let's remove commented code and add it later if needed please :).

File: lib/assets/test/spec/cartodb3/editor/layers/basemap-content-views/basemap-categories-view.spec.js [hunk 2]
```diff
@@ -38,13 +38,13 @@ describe('editor/layers/basemap-content-views/basemap-categories-view', function
...
         return 'Color';
       }
-    }, {
-      selected: false,
-      val: 'Mapbox',
-      label: 'Mapbox',
-      template: function () {
-        return 'Mapbox';
-      }
+    // }, {
```
Comment (xavijam) [inline lib/assets/test/spec/cartodb3/editor/layers/basemap-content-views/basemap-categories-view.spec.js]: Same here

File: lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-select-view.js [hunk 3]
```diff
@@ -32,6 +37,12 @@ module.exports = CoreView.extend({
       case 'color':
         this._renderForm();
         break;
+      case 'custom':
+        this._renderCustomMosaic();
+        break;
+      // case 'mapbox':
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-select-view.js]: Dixit!

File: lib/assets/javascripts/cartodb3/components/modals/add-basemap/xyz/xyz-view.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    var urlErrorMsg;
+
+    var url = this.$('.js-url').val();
+    var tms = this.model.get('tms');
+
+    if (url) {
+      try {
+        layer = this._byCustomURL(url, tms);
+      } catch (e) {
+        urlErrorMsg = 'It does not look like a valid XYZ URL';
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/modals/add-basemap/xyz/xyz-view.js]: Translate string please

File: lib/assets/javascripts/cartodb3/components/modals/add-basemap/xyz/xyz-view.js [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            self._indicateIsValidating(false);
+            self._updateError();
+          }
+        },
+        error: function () {
+          if (thisCallSeq === self._lastCallSeq) {
+            self._disableOkBtn(false);
+            self._indicateIsValidating(false);
+            // Note that this text can not be longer, or it will exceed available space of the error label.
+            self._updateError("We couldn't validate this, if you're sure it contains data click \"add basemap\"");
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/modals/add-basemap/xyz/xyz-view.js]: Translate string please

File: lib/assets/javascripts/cartodb3/components/modals/add-basemap/xyz/xyz-view.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+  _disableOkBtn: function (disable) {
+    this.$('.ok')[ disable ? 'addClass' : 'removeClass' ]('is-disabled');
+  },
+
+  _updateError: function (msg) {
+    this.$('.js-error').text(msg)[ msg ? 'addClass' : 'removeClass' ]('is-visible');
+  },
+
+  _indicateIsValidating: function (indicate) {
+    if (indicate) {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/modals/add-basemap/xyz/xyz-view.js]: `this.$('.js-validating').toggle(!!indicate);`

File: lib/assets/javascripts/cartodb3/components/modals/add-basemap/xyz/xyz-view.js [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    }
+  },
+
+  _changeTMS: function (e) {
+    this.model.set('tms', !this.model.get('tms'));
+    this._onKeydown(e);
+  },
+
+  _setTMSCheckbox: function (e) {
+    this.$('.js-tms .Checkbox-input')[ this.model.get('tms') ? 'addClass' : 'removeClass' ]('is-checked');
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/modals/add-basemap/xyz/xyz-view.js]: this.$('.js-tms .Checkbox-input').toggleClass('is-checked', this.model.get('tms'));

File: lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js [hunk 8]
```diff
@@ -237,14 +237,14 @@ module.exports = Backbone.Collection.extend({
     var me = this.getBaseLayer().toJSON();
     var other = layer;
 
-    var myType = me.type ? me.type : me.options.type;
-    var itsType = other.type ? other.type : other.options.type;
+    var myType = (me.type !== undefined) ? me.type : me.options.type;
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: This part is a bit weird, doing the same for each variable, let me think in a pattern...

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: different that undefined and null maybe?
`me.whatever != null ? _ : _;` could do the trick.

Comment (matallo) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: yeah, it's _inspired_ mostly in https://github.com/CartoDB/cartodb.js/blob/develop/src/geo/map.js#L20

let me know of a better way to do this, happy to improve it!

File: lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-mosaic-add-item-view.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+  _onClick: function () {
+    var self = this;
+
+    var modal = this._modals.create(function (modalModel) {
+      var addBasemapModel = new AddBasemapModel({}, {
+        layerDefinitionsCollection: self._layerDefinitionsCollection,
+        basemapsCollection: self._basemapsCollection,
+        customBaselayersCollection: self._customBaselayersCollection
+        // currentTab: self._currentTab
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-mosaic-add-item-view.js]: Not needed anymore?

File: app/assets/stylesheets/common/form-content.css.scss [hunk 10]
```diff
@@ -283,6 +283,9 @@ $sLabel-width: 140px;
   border-radius: 4px;
   @include transition(margin, 200ms, ease-in-out);
 }
+.Form-inputError.Form-inputError--small {
```
Comment (xavijam) [inline app/assets/stylesheets/common/form-content.css.scss]: Uhm... I would not add this in the old stylesheets, we have the `CDB-Size` class for this purpose in CartoAssets.

File: lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-mosaic-add-item-view.js (hunk 9 above)
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-mosaic-add-item-view.js]: ready for next tabs, when opening from a different category will open that tab

**Comment by matallo** [comment]
made the implementation for fallback thumbnails both in layer list and mosaic, currently it's an empty image, but they could be replaced once @josecruz pass the design

so this was the last point for review @xavijam 

screen shot 2016-08-11 at 14 29 29
screen shot 2016-08-11 at 14 30 14

**Comment by matallo** [comment]
close https://github.com/CartoDB/cartodb/issues/9396

**Pull request merged**